S/StableHLO/build_tarballs.jl

```diff
@@ -0,0 +1,111 @@
+# Note that this script can accept some limited command-line arguments, run
+# `julia build_tarballs.jl --help` to see a usage message.
+using BinaryBuilder, Pkg
+
+const YGGDRASIL_DIR = "../.."
+include(joinpath(YGGDRASIL_DIR, "fancy_toys.jl"))
+include(joinpath(YGGDRASIL_DIR, "platforms", "llvm.jl"))
+
+name = "StableHLO"
+version = v"0.14.6"
+
+sources = [
+    GitSource("https://github.com/openxla/stablehlo.git", "8816d0581d9a5fb7d212affef858e991a349ad6b"),
+]
+
+llvm_versions = [v"17.0.6+0"]
+
+script = raw"""
+cd $WORKSPACE/srcdir/stablehlo
+
+# need to run mlir-tblgen and mlir-pdll on the host
+rm ${prefix}/bin/mlir-tblgen ${prefix}/bin/mlir-pdll
+ln -s ${host_prefix}/bin/mlir-tblgen ${prefix}/bin/mlir-tblgen
+ln -s ${host_prefix}/bin/mlir-pdll ${prefix}/bin/mlir-pdll
+
+CMAKE_FLAGS=()
+CMAKE_FLAGS+=(-DCMAKE_INSTALL_PREFIX=${prefix})
+CMAKE_FLAGS+=(-DCMAKE_BUILD_TYPE=Release)
+CMAKE_FLAGS+=(-DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN})
+CMAKE_FLAGS+=(-DCMAKE_CROSSCOMPILING:BOOL=ON)
+
+CMAKE_FLAGS+=(-DLLVM_DIR=${prefix}/lib/cmake/llvm)
+CMAKE_FLAGS+=(-DLLVM_ENABLE_ASSERTIONS=OFF)
+CMAKE_FLAGS+=(-DLLVM_LINK_LLVM_DYLIB=ON)
+CMAKE_FLAGS+=(-DLLVM_ENABLE_LLD="OFF")
+
+CMAKE_FLAGS+=(-DMLIR_DIR=${prefix}/lib/cmake/mlir)
+
+# building shared libraries crashes on linking, so build static libraries and link them manually
+CMAKE_FLAGS+=(-DBUILD_SHARED_LIBS=OFF)
+
+cmake -B build -S . -GNinja ${CMAKE_FLAGS[@]}
+ninja -C build -j 1 install
```

Comment:
Why `-j1`?

Reply:
Looks like there are race conditions when running multithread. I ran it in just one thread for debugging but it just worked.

Reply:
This is the error I get if I do `-j ${nproc}`
```
[16:43:12] In file included from /workspace/srcdir/stablehlo/stablehlo/conversions/tosa/transforms/StablehloPrepareForTosa.cpp:23:
[16:43:12] /workspace/srcdir/stablehlo/stablehlo/dialect/StablehloOps.h:45:10: fatal error: stablehlo/dialect/StablehloEnums.h.inc: No such file or directory
[16:43:12]    45 | #include "stablehlo/dialect/StablehloEnums.h.inc"
[16:43:12]       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[16:43:12] compilation terminated.
[16:43:12] ninja: job failed: /opt/bin/x86_64-linux-gnu-libgfortran5-cxx11/x86_64-linux-gnu-g++ --sysroot=/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/ -DGTEST_HAS_RTTI=0 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/workspace/srcdir/stablehlo -I/workspace/srcdir/stablehlo/build -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -Wimplicit-fallthrough -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG   -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS  -fno-exceptions -funwind-tables -fno-rtti -std=gnu++17 -MD -MT stablehlo/conversions/tosa/transforms/CMakeFiles/obj.StablehloTOSATransforms.dir/StablehloPrepareForTosa.cpp.o -MF stablehlo/conversions/tosa/transforms/CMakeFiles/obj.StablehloTOSATransforms.dir/StablehloPrepareForTosa.cpp.o.d -o stablehlo/conversions/tosa/transforms/CMakeFiles/obj.StablehloTOSATransforms.dir/StablehloPrepareForTosa.cpp.o -c /workspace/srcdir/stablehlo/stablehlo/conversions/tosa/transforms/StablehloPrepareForTosa.cpp
[16:43:12] In file included from /workspace/srcdir/stablehlo/stablehlo/conversions/tosa/transforms/StablehloLegalizeToTosa.cpp:25:
[16:43:12] /workspace/srcdir/stablehlo/stablehlo/dialect/StablehloOps.h:45:10: fatal error: stablehlo/dialect/StablehloEnums.h.inc: No such file or directory
[16:43:12]    45 | #include "stablehlo/dialect/StablehloEnums.h.inc"
[16:43:12]       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[16:43:12] compilation terminated.
[16:43:12] ninja: job failed: /opt/bin/x86_64-linux-gnu-libgfortran5-cxx11/x86_64-linux-gnu-g++ --sysroot=/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/ -DGTEST_HAS_RTTI=0 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/workspace/srcdir/stablehlo -I/workspace/srcdir/stablehlo/build -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -Wimplicit-fallthrough -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG   -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS  -fno-exceptions -funwind-tables -fno-rtti -std=gnu++17 -MD -MT stablehlo/conversions/tosa/transforms/CMakeFiles/obj.StablehloTOSATransforms.dir/StablehloLegalizeToTosa.cpp.o -MF stablehlo/conversions/tosa/transforms/CMakeFiles/obj.StablehloTOSATransforms.dir/StablehloLegalizeToTosa.cpp.o.d -o stablehlo/conversions/tosa/transforms/CMakeFiles/obj.StablehloTOSATransforms.dir/StablehloLegalizeToTosa.cpp.o -c /workspace/srcdir/stablehlo/stablehlo/conversions/tosa/transforms/StablehloLegalizeToTosa.cpp
[16:43:15] ninja: subcommands failed
```

Reply:
I guess that this was solved in more recent versions. This is an old version, but it is the last one supporting LLVM 17. The newest versions (currently I think they are in 0.19, so 5 minor versions above) use commits which are only present in yet unreleased LLVM 18.